File: c++/leetcode/1089. Duplicate Zeros.cpp
class Solution
{
public:
  void duplicateZeros(vector<int> &arr)
  {
    int end = arr.size();
    int j = end + count(arr.begin(), arr.end(), 0);

    for (int i = end - 1; i >= 0; --i)
    {
      if (--j < end)
        arr[j] = arr[i];

      if (arr[i] == 0 && --j < end)
        arr[j] = 0;
    }
  }

File: javascript/leetcode/1089. Duplicate Zeros.js
/**
 * @param {number[]} arr
 * @return {void} Do not return anything, modify arr in-place instead.
 */
var duplicateZeros = function (arr) {
  let end = arr.length,
    j = end;

  arr.forEach((el) => {
    if (!el) ++j;
  });

  for (let i = end - 1; i >= 0; --i) {
    if (--j < end) arr[j] = arr[i];

    if (arr[i] === 0 && --j < end) arr[j] = 0;
  }
};
